### PR TITLE
New arch

### PR DESCRIPTION
**Pull request by sawal-timsina**
### folder structor changed
- [x] seperated CSR into domain
- [x] moved infrastructure to internal/config
- [x] seperated external services
- [x] Moved migrations,models and seeds to database folder

### Refactors
- [x] Response and errors
- [x] Seperated DB dsn
- [x] Seperated Env Path: usable for tests

## Upcomming Changes
- [x] Refactor JWT Auth Service and Controller
- [x] Refactor Validator (single validator for all requests)
- [x] Refactor api_error
- [x] Memory optimization (dependencies) - High Priority
- [ ] Create CLI app and Move Makefile commands
- [x] Move Business logics to Domain specific service
- [ ] Refactor Tests 
- [x] pre-commit and lints
- [x] module based logger
- [x] create plug&play packages of services we use
- [ ] Add Mocks data - High Priority
- [ ] Allow to feed Mocks data from cli - High Priority
- [x] Seperate API based on actor (admin, user)
- [x] Change migration package - [Task](https://readytowork.atlassian.net/browse/BD-210) | Ashish Dhakal

### CODE REVIEW

**Review by Simant-Thapa-Magar — COMMENTED** [review]
1. Image is not being built as go version seems different on different files. Error message says version 1.23 is required
2. Commands seem to be executing based on old structure. For eg:
- make create is creating migration files based on old structure
- make crud is throwing error of file not found
Please make sure each commands used are functioning as per changed structure

**Comment by sawal-timsina** [comment]
> 1. Image is not being built as go version seems different on different files. Error message says version 1.23 is required
> 2. Commands seem to be executing based on old structure. For eg:
> 
> * make create is creating migration files based on old structure
> * make crud is throwing error of file not found
>   Please make sure each commands used are functioning as per changed structure

@Simant-Thapa-Magar  Thank you for taking your time to review this. I am quite busy in GS, would you mind taking a look at it if you are free?

**Comment by Simant-Thapa-Magar** [comment]
> > 1. Image is not being built as go version seems different on different files. Error message says version 1.23 is required
> > 2. Commands seem to be executing based on old structure. For eg:
> > 
> > 
> > * make create is creating migration files based on old structure
> > * make crud is throwing error of file not found
> >   Please make sure each commands used are functioning as per changed structure
> 
> @Simant-Thapa-Magar Thank you for taking your time to review this. I am quite busy in GS, would you mind taking a look at it if you are free?

Apologies, I am busy as well with easyy-kids & easyy-adults. Please take your time as per convenience 🙇

**Comment by sawal-timsina** [comment]
@smooth-55 thank you for the changes

File: services/stripe_test.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+	//}
+
+	//stripeClient := client.New("sk_test", stripeTestBackends)
+
+	env := config.NewEnv(
+		".env",
+	)
+	stripeService := NewStripeService(
+		env,
+		config.GetLogger(env),
```
Comment (smooth-55) [inline services/stripe_test.go]: config.GetLogger expects no arguments. Getting err `too many arguments in call to config.GetLogger
	have ("boilerplate-api/internal/config".Env)
	want ()`

File: services/stripe_test.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
```
Comment (smooth-55) [inline services/stripe_test.go]: Here too

File: services/stripe_test.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	//	Uploads: stripeBackendMock,
+	//}
+
+	//stripeClient := client.New("sk_test", stripeTestBackends)
+	env := config.NewEnv(
+		".env",
+	)
+	stripeService := NewStripeService(
+		env,
+		config.GetLogger(env),
```
Comment (smooth-55) [inline services/stripe_test.go]: Here too

File: services/stripe_test.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	//}
+
+	//stripeClient := client.New("sk_test", stripeTestBackends)
+
+	env := config.NewEnv(
+		".env",
+	)
+	stripeService := NewStripeService(
+		env,
+		config.GetLogger(env),
```
Comment (smooth-55) [inline services/stripe_test.go]: Here too

File: tests/controllers_i_test/controllers_test.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
+package controllers_i_test
+
+import (
+	"boilerplate-api/api/controllers"
+	"boilerplate-api/api/repository"
+	"boilerplate-api/api/services"
```
Comment (smooth-55) [inline tests/controllers_i_test/controllers_test.go]: There's no such package `controllers, repository, services` to import.
getting err:
`could not import boilerplate-api/api/controllers (no required module provides package "boilerplate-api/api/controllers")`

File: tests/repository_test/repository_test.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
+package repository_test
+
+import (
+	"boilerplate-api/api/repository"
```
Comment (smooth-55) [inline tests/repository_test/repository_test.go]: could not import boilerplate-api/api/repository (no required module provides package "boilerplate-api/api/repository")

**Review by smooth-55 — DISMISSED** [review]

**Review by smooth-55 — COMMENTED** [review]
@sawal-timsina I've created two entity `user` and `product` and when I tried to call `getAllProductsByUserId` service in `user` package's controller and `GetOneUser` service in product package's controller I am getting import cycle issue.
Reason behind this issue is, this architecture is using same package for controllers, service, repository and routes.
we'll need to create separate package for each entities controllers, services, repository and routes.
like, user_service, user_repository and user_routes.

Screenshot 2024-05-20 at 5 15 30 PM

===
Also, api endpoint is grouped by version, like api/v1/user. As per my perspective we also need to separate entities based on version itself as below.

Screenshot 2024-05-20 at 5 17 25 PM


cc: @Simant-Thapa-Magar

**Comment by sawal-timsina** [comment]
> @sawal-timsina I've created two entity `user` and `product` and when I tried to call `getAllProductsByUserId` service in `user` package's controller and `GetOneUser` service in product package's controller I am getting import cycle issue. Reason behind this issue is, this architecture is using same package for controllers, service, repository and routes. we'll need to create separate package for each entities controllers, services, repository and routes. like, user_service, user_repository and user_routes.
> 
> Screenshot 2024-05-20 at 5 15 30 PM
> === Also, api endpoint is grouped by version, like api/v1/user. As per my perspective we also need to separate entities based on version itself as below.
> 
> Screenshot 2024-05-20 at 5 17 25 PM
> cc: @Simant-Thapa-Magar

@smooth-55 What do you suggest? How should we fix this?

**Comment by smooth-55** [comment]
> > @sawal-timsina I've created two entity `user` and `product` and when I tried to call `getAllProductsByUserId` service in `user` package's controller and `GetOneUser` service in product package's controller I am getting import cycle issue. Reason behind this issue is, this architecture is using same package for controllers, service, repository and routes. we'll need to create separate package for each entities controllers, services, repository and routes. like, user_service, user_repository and user_routes.
> > Screenshot 2024-05-20 at 5 15 30 PM
> > === Also, api endpoint is grouped by version, like api/v1/user. As per my perspective we also need to separate entities based on version itself as below.
> > Screenshot 2024-05-20 at 5 17 25 PM
> > cc: @Simant-Thapa-Magar
> 
> @smooth-55 What do you suggest? How should we fix this?

we'll need to create separate package for each entities controllers, services, repository and routes. like, user_service, user_repository and user_routes.

or we can do same as [THIS REPO](https://github.com/readytowork-org/go-gin-skeleton-app-based)

**Pull request merged**